### ISSUE TITLE
Make bolt regression network

### ISSUE THREAD

**Issue by benitoThree** [ThirdAILabs/Universe#227]
Parent: #225 

### Problem
Currently, bolt only does classification tasks. To run regression tasks, we can take the existing classification network, remove the activation function of the final layer, and switch the loss function to MeanSquaredError loss function. However, if we do sparse training followed by dense inference, predictions get blown up by 1/load_factor (e.g. load_factor=0.01 -> predictions scaled by 100x).

### Final solution
Sparse inference

### Solutions / TODO
- [x] Train with two output neurons y1 and y2, then train to fit y1 to truth value and fit y2 to 1; train to fit (y1, y2) to (truth, 1). When doing inference, the predicted value is x1/x2.
    * Promising at first but doesn't actually work because y1 and y2 are optimized independently. Thus, y1 is prioritized in some examples and y2 is prioritized in others. In reality, the model cannot properly learn to fit y2 to 1 while optimizing for y1 at the same time. Thus, we get scenarios where y2 gets fit to 0.3 or 2 instead of 1 and it wouldn't work. Solution 2 should solve this, though.
    * Would be interesting to try to store the average y2 during training and scale it by this amount.
    * Interesting observation: y1/y2 always equals the same number when network is dense. When sparse, y1/y2 varies per example. y1/y2 is intended to fit to truth value. Since each example has a different truth value, y1/y2 should be different per example, as in the sparse network case. However, since y1/y2 is the same across examples when the network is dense, and because the variance of y1/y2 in sparse network does not correlate well with the actual truth values, it seems the variance is because each example activates a different set of neurons.
    * Changed activation to leaky relu. If run for 20 epochs, seems to stop giving the same ratio.
    * Somewhat works if dense training.
- [x] Train with two output neurons y1 and y2, then train to fit y1/y2 to truth value. When doing inference, the predicted value is y1/y2.
    * For some reason this also keeps giving the same ratio, and stays this way even after running many epochs. Also, the ratio does not converge to the target for some reason. Seems like a bug. Currently investigating.
    * This works after changing activation function to relu + 1 (bias, so x2 never reaches 0) and training is dense
    * Unsurprisingly this doesn't work when doing sparse training. This hasn't been verified but we can imagine x2's value before activation during sparse training to be way in the negatives but is 1 after activation, such that even when we activate all the neurons, x2 is still in the negatives before activation, and gives 1, therefore it is useless for weighting
    * Want to try using softmax activation. Doesn't seem to work either... x1/x2 is very very very low... Or nan
    * Optimize for Manhattan distance instead? Doesn't change much.
- [x] Sparse inference
- [x] Just divide by sparsity
    * Take the product of the sparsities of all layers in the network, store this, and multiply with the inference output. This works great so far. Tested with some evenly distributed mock datasets and the featured-extracted demand forecasting dataset and we're getting good RMSE.
    - [ ] Put tests in one file. As a metric, make sure that inference error < 110% of training error
        - [x] run on trivial mock dataset
        - [x] run on mock dataset that is curated to have unbalanced clusters. 
        - [x] run on mock dataset that will force each sample to choose a subnetwork. One way to do this is to have a number of cosine distance-separable (SRP separable) clusters that have independent underlying input->output functions.
            * This dataset destroyed this approach. Training accuracy ~ 20 rmse, testing accuracy ~ 200 rmse.
        - [x] run on featurized demand forecasting dataset from kaggle
    * Implement:
        - [ ] WMAPE
        - [x] bolt regression network
            * train and predict return epoch time, loss, and metric
            * write loss and metric functions (instead of hard coding into network)
    * Cool idea:
        - [ ] Use a linear model to learn multiplying constant given neuron activation distribution.
- [ ] Make new activation function that divides by sparsity
- [x] Train to predict x1 = truth and x2 = truth + C. Then multiply x1 by (x2 - x1) / C
    * This doesn't work since C will only affect the bias, and the bias isn't affected by previous layers.

### Questions / roadblocks
* How do I enable sparsity during inference? Called the .use_sparse_inference() method, but it doesn't seem to work; as a sanity check, I ran training and inference with the same data. Theoretically, I should get the same RMSE value for both, but they came out different.
* Once we figure out a working method, how should we ,
